File: python/tests/wire/vbp/test_frame.py
"""Tests for the vbp.wire.frame module — frame encode/decode round-trips."""
from __future__ import annotations

import io
import os
import struct
import sys
import unittest

# Make sure we can import the vedadb package from the repo root.
HERE = os.path.dirname(os.path.abspath(__file__))
REPO_ROOT = os.path.abspath(os.path.join(HERE, "..", "..", ".."))
if REPO_ROOT not in sys.path:
    sys.path.insert(0, REPO_ROOT)

from vedadb.wire.vbp.frame import (  # noqa: E402
    Frame,
    MAX_FRAME_LEN,
    VBPBadMagic,
    VBPConnectionClosed,
    VBPFrameTooLarge,
    VBPFrameTooShort,
    VBPProtocolError,
    frame_bytes,
    read_frame,
    read_frame_bytes,
    write_frame,
)


class TestFrameRoundTrip(unittest.TestCase):
    """A written frame is decodable back to the same Frame."""

    def test_empty_body(self):
        data = frame_bytes(seq=0, op=0x01, flags=0, body=b"")
        f, n = read_frame_bytes(data)
        self.assertEqual(f.seq, 0)
        self.assertEqual(f.op, 0x01)
        self.assertEqual(f.flags, 0)
        self.assertEqual(f.body, b"")
        self.assertEqual(n, len(data))
        self.assertEqual(n, 8 + 2)  # 8-byte header + 2-byte op+flags

    def test_short_body(self):
        body = b"\x01\x02\x03"
        data = frame_bytes(seq=42, op=0x06, flags=0x01, body=body)
        f, n = read_frame_bytes(data)
        self.assertEqual(f.seq, 42)
        self.assertEqual(f.op, 0x06)
        self.assertEqual(f.flags, 0x01)
        self.assertEqual(f.body, body)
        self.assertEqual(n, len(data))

    def test_seq_wraps_at_256(self):
        # The wire uses a 1-byte seq, so values >= 256 are masked to fit.
        # We expect frame_bytes to either reject (raise) or mask — the
        # spec just requires the on-wire value to match (seq mod 256).
        # The v1 implementation raises ValueError for seq > 0xFF
        # (callers should allocate seqs via Multiplexer, not pass 256+).
        # This test documents that behavior.
        for seq in (0, 1, 127, 255):
            data = frame_bytes(seq=seq, op=0x16, flags=0, body=b"")
            f, _ = read_frame_bytes(data)
            self.assertEqual(f.seq, seq)
        # For seq > 0xFF, the Multiplexer's _alloc_seq masks to a byte
        # (the seq is `(self._next_seq + 1) & 0xFF`).  Direct frame_bytes
        # callers get a ValueError as a safety net.
        with self.assertRaises(ValueError):
            frame_bytes(seq=256, op=0x16, flags=0, body=b"")

    def test_magic_is_VDB(self):
        data = frame_bytes(seq=0, op=0x01, flags=0, body=b"")
        self.assertEqual(data[:3], b"VDB")

    def test_length_is_le_u32(self):
        body = b"\x00" * 100
        data = frame_bytes(seq=0, op=0x01, flags=0, body=body)
        # payload_length = op + flags + body = 2 + 100 = 102
        (plen,) = struct.unpack("<I", data[3:7])
        self.assertEqual(plen, 102)

    def test_large_body(self):
        body = b"x" * 4096
        data = frame_bytes(seq=0, op=0x06, flags=0, body=body)
        f, n = read_frame_bytes(data)
        self.assertEqual(f.body, body)
        self.assertEqual(n, len(data))


class TestFrameErrors(unittest.TestCase):
    """Malformed frames are rejected with the right exception class."""

    def test_bad_magic(self):
        # 'XYZ' instead of 'VDB'
        data = b"XYZ" + struct.pack("<I", 2) + b"\x00\x00"
        with self.assertRaises(VBPBadMagic):
            read_frame_bytes(data)

    def test_payload_too_short(self):
        # payload_length = 1 (less than op+flags minimum of 2)
        data = b"VDB" + struct.pack("<I", 1) + b"\x00"
        with self.assertRaises(VBPFrameTooShort):
            read_frame_bytes(data)

    def test_payload_too_large(self):
        # payload_length = MAX_FRAME_LEN + 1
        data = b"VDB" + struct.pack("<I", MAX_FRAME_LEN + 1) + b"\x00\x00"
        with self.assertRaises(VBPFrameTooLarge):
            read_frame_bytes(data)

    def test_incomplete_data_raises(self):
        # Only 5 bytes — not enough for the 8-byte header.
        with self.assertRaises(ValueError):
            read_frame_bytes(b"VDB\x00\x00")

    def test_incomplete_body_raises(self):
        # Header says 100 bytes body, only 50 follow.
        data = b"VDB" + struct.pack("<I", 102) + b"\x00\x00" + b"x" * 50
        with self.assertRaises(ValueError):
            read_frame_bytes(data)


class TestStreamRoundTrip(unittest.TestCase):
    """A frame written to a stream is read back the same."""

    def test_write_then_read(self):
        body = b"hello, world"
        buf = io.BytesIO()
        write_frame(buf, seq=7, op=0x05, flags=0, body=body)
        buf.seek(0)
        f = read_frame(buf)
        self.assertEqual(f.seq, 7)
        self.assertEqual(f.op, 0x05)
        self.assertEqual(f.body, body)

    def test_write_close_then_read(self):
        """Closing the write side doesn't break read."""
        body = b"abc"
        buf = io.BytesIO()
        write_frame(buf, 0, 0x16, 0, body)
        buf.seek(0)
        f = read_frame(buf)
        self.assertEqual(f.body, body)

    def test_read_past_eof_raises(self):
        buf = io.BytesIO(b"VDB\x00")  # truncated header
        with self.assertRaises(VBPProtocolError):
            read_frame(buf)


class TestFrameEquality(unittest.TestCase):
    """Two frames with the same fields are equal; different fields differ."""

    def test_equal(self):
        a = Frame(seq=1, op=2, flags=3, body=b"x")
        b = Frame(seq=1, op=2, flags=3, body=b"x")
        self.assertEqual(a, b)

    def test_not_equal(self):
        a = Frame(seq=1, op=2, flags=3, body=b"x")
        b = Frame(seq=2, op=2, flags=3, body=b"x")
        self.assertNotEqual(a, b)

    def test_repr(self):
        f = Frame(seq=0, op=0x01, flags=0, body=b"")
        self.assertIn("Frame", repr(f))
        self.assertIn("seq=0", repr(f))
        self.assertIn("op=0x01", repr(f))


class TestValidation(unittest.TestCase):
    """Frame constructor and write_frame reject out-of-range values."""

    def test_ctor_rejects_seq(self):
        with self.assertRaises(ValueError):
            Frame(seq=256, op=0, flags=0, body=b"")

    def test_ctor_rejects_op(self):
        with self.assertRaises(ValueError):
            Frame(seq=0, op=256, flags=0, body=b"")

    def test_ctor_rejects_flags(self):
        with self.assertRaises(ValueError):
            Frame(seq=0, op=0, flags=256, body=b"")

    def test_write_rejects_oversize(self):
        with self.assertRaises(VBPFrameTooLarge):
            frame_bytes(0, 0, 0, b"x" * (MAX_FRAME_LEN + 1))


if __name__ == "__main__":
    unittest.main()

File: node/tests/wire/vbp/test_frame.js
'use strict';

const { test } = require('node:test');
const assert = require('node:assert');
const {
  Frame, frameBytes, writeFrame, tryDecodeFrame, feedBytes, resetState,
  VBPProtocolError, VBPBadMagic, VBPFrameTooShort, VBPFrameTooLarge,
  VBPConnectionClosed, MAGIC, HDR_LEN, OPFLAGS_LEN, MAX_FRAME_LEN, DEFAULT_VBP_PORT,
} = require('../../../src/wire/vbp/frame');

function makeState() { return { buf: Buffer.alloc(0), offset: 0 }; }

test('frame: encode/decode round-trip', () => {
  const body = Buffer.from('hello world');
  const bytes = frameBytes(0x42, 0x06, 0x00, body);
  // 3 + 4 + 1 + 1 + 1 + 11 = 21
  assert.strictEqual(bytes.length, 21);
  assert.deepStrictEqual(bytes.subarray(0, 3), MAGIC);
  assert.strictEqual(bytes.readUInt32LE(3), 2 + body.length);
  assert.strictEqual(bytes[7], 0x42);
  assert.strictEqual(bytes[8], 0x06);
  assert.strictEqual(bytes[9], 0x00);
  assert.deepStrictEqual(bytes.subarray(10), body);

  const state = makeState();
  feedBytes(state, bytes);
  const f = tryDecodeFrame(state);
  assert.ok(f instanceof Frame);
  assert.strictEqual(f.seq, 0x42);
  assert.strictEqual(f.op, 0x06);
  assert.strictEqual(f.flags, 0);
  assert.deepStrictEqual(f.body, body);
});

test('frame: empty body', () => {
  const bytes = frameBytes(0x01, 0x0C, 0x00, Buffer.alloc(0));
  assert.strictEqual(bytes.length, 10);
  const state = makeState();
  feedBytes(state, bytes);
  const f = tryDecodeFrame(state);
  assert.strictEqual(f.body.length, 0);
  assert.strictEqual(f.op, 0x0C);
});

test('frame: seq 0 boundary', () => {
  const f = new Frame(0, 0x01, 0, Buffer.alloc(0));
  assert.strictEqual(f.seq, 0);
});

test('frame: out-of-range seq throws', () => {
  assert.throws(() => new Frame(-1, 0x01, 0, Buffer.alloc(0)), RangeError);
  assert.throws(() => new Frame(256, 0x01, 0, Buffer.alloc(0)), RangeError);
});

test('frame: out-of-range op throws', () => {
  assert.throws(() => new Frame(1, 256, 0, Buffer.alloc(0)), RangeError);
  assert.throws(() => new Frame(1, -1, 0, Buffer.alloc(0)), RangeError);
});

test('frame: out-of-range flags throws', () => {
  assert.throws(() => new Frame(1, 0x01, 256, Buffer.alloc(0)), RangeError);
  assert.throws(() => new Frame(1, 0x01, -1, Buffer.alloc(0)), RangeError);
});

test('frame: bad magic throws VBPBadMagic', () => {
  const state = makeState();
  const bytes = Buffer.from('XYZ\x05\x00\x00\x00\x01\x02\x03');
  feedBytes(state, bytes);
  assert.throws(() => tryDecodeFrame(state), VBPBadMagic);
});

test('frame: payload too short throws VBPFrameTooShort', () => {
  const state = makeState();
  const bytes = Buffer.concat([MAGIC, Buffer.from([0x01, 0x00, 0x00, 0x00]), Buffer.from([0x01, 0x02, 0x03])]);
  feedBytes(state, bytes);
  // payload_len = 1, < 2 (OPFLAGS_LEN)
  assert.throws(() => tryDecodeFrame(state), VBPFrameTooShort);
});

test('frame: payload too large throws VBPFrameTooLarge', () => {
  const state = makeState();
  const tooBig = Buffer.alloc(8);
  MAGIC.copy(tooBig, 0);
  tooBig.writeUInt32LE(MAX_FRAME_LEN + 1, 3);
  tooBig[7] = 0x01;
  feedBytes(state, tooBig);
  assert.throws(() => tryDecodeFrame(state), VBPFrameTooLarge);
});

test('frame: incomplete frame returns null (no throw)', () => {
  const state = makeState();
  const bytes = frameBytes(0x05, 0x06, 0, Buffer.from('abc'));
  // Feed only half.
  feedBytes(state, bytes.subarray(0, 12));
  assert.strictEqual(tryDecodeFrame(state), null);
  // Feed the rest.
  feedBytes(state, bytes.subarray(12));
  const f = tryDecodeFrame(state);
  assert.ok(f);
  assert.strictEqual(f.seq, 0x05);
});

test('frame: multiple frames in one chunk', () => {
  const a = frameBytes(0x01, 0x06, 0, Buffer.from('hello'));
  const b = frameBytes(0x02, 0x07, 0, Buffer.from('world'));
  const state = makeState();
  feedBytes(state, Buffer.concat([a, b]));
  const fa = tryDecodeFrame(state);
  const fb = tryDecodeFrame(state);
  assert.strictEqual(fa.seq, 0x01);
  assert.strictEqual(fb.seq, 0x02);
  assert.deepStrictEqual(fa.body, Buffer.from('hello'));
  assert.deepStrictEqual(fb.body, Buffer.from('world'));
});

test('frame: writeFrame appends to a sink', () => {
  const sink = [];
  const buf = frameBytes(0x03, 0x08, 0, Buffer.from('xyz'));
  sink.push(buf);
  // Re-encode through the writer.
  const out = Buffer.concat(sink);
  const state = makeState();
  feedBytes(state, out);
  const f = tryDecodeFrame(state);
  assert.strictEqual(f.op, 0x08);
  assert.deepStrictEqual(f.body, Buffer.from('xyz'));
});

test('frame: 64-byte payload boundary', () => {
  const body = Buffer.alloc(64, 0xAA);
  const bytes = frameBytes(0x10, 0x06, 0, body);
  const state = makeState();
  feedBytes(state, bytes);
  const f = tryDecodeFrame(state);
  assert.strictEqual(f.body.length, 64);
});

test('frame: default VBP port is 6380', () => {
  assert.strictEqual(DEFAULT_VBP_PORT, 6380);
});

test('frame: max frame is 64 MiB', () => {
  assert.strictEqual(MAX_FRAME_LEN, 64 * 1024 * 1024);
});

test('frame: HDR_LEN is 8 (3 magic + 4 len + 1 seq)', () => {
  assert.strictEqual(HDR_LEN, 8);
});

test('frame: OPFLAGS_LEN is 2 (op + flags)', () => {
  assert.strictEqual(OPFLAGS_LEN, 2);
});
